Include/Rover/ScalarView.hpp

```diff
@@ -0,0 +1,179 @@
+#ifndef ROVER_SCALAR_VIEW_HPP
+#define ROVER_SCALAR_VIEW_HPP
+#include <functional>
+#include <optional>
+#include <vector>
+
+namespace Rover {
+
+  //! Interface to pass adapted Trial Samples to typed Algorithms.
+  /*!
+    \tparam T CompType used by the algorithm.
+  */
+  template<typename T>
+  class ScalarView {
+    public:
+
+      //! CompType used by the algorithm.
+      using Type = T;
+
+      //! Adapted sample.
+      struct Sample {
+
+        //! The type of the result.
+        using Result = Type;
+
+        //! The type of the arguments.
+        using Arguments = std::vector<Type>;
+
+        //! The result of the trial's function.
+        Result m_result;
+
+        //! The arguments passed to the trial's function.
+        Arguments m_arguments;
+      };
+
+      //! Constant input iterator.
+      class ConstIterator {
+        public:
+          
+          //! Checks whether two iterators point to the same Sample.
+          bool operator ==(ConstIterator other) const;
+
+          //! Checks whether two iterators do not point to the same Sample.
+          bool operator !=(ConstIterator other) const;
+
+          //! Returns a reference to the Sample the iterator points to.
+          const Sample& operator *() const;
+
+          //! Returns a pointer to the Sample the iterator points to.
+          const Sample* operator ->() const;
+
+          //! Increments the iterator.
+          ConstIterator& operator++();
+
+          //! Increments the iterator.
+          ConstIterator operator++(int);
+
+        private:
+          friend class ScalarView;
+
+          ConstIterator(const ScalarView* reader, std::size_t index);
+          std::optional<Sample> retrieve_sample() const;
+
+          const ScalarView* m_reader;
+          std::size_t m_index;
+          std::optional<Sample> m_sample;
+      };
+
+      //! Creates a ScalarView.
+      /*!
+        \param get The function taking an index i and returning the ith adapted
+                   sample.
+        \param size The total number of samples.
+      */
+      template<typename GetFwd>
+      ScalarView(GetFwd&& get, std::size_t size);
+
+      //! Returns an input iterator to the beginning of the trial.
+      ConstIterator begin() const;
+
+      //! Returns an input iterator to the end of the trial.
+      ConstIterator end() const;
+
+      //! Returns ith adapted sample.
+      Sample operator [](std::size_t i) const;
+
+      //! Returns the number of the samples.
+      std::size_t size() const;
+
+    private:
+      std::function<Sample (std::size_t)> m_get;
```

Comment:
Remove the use of ```std::function```. Add another template parameter if need be.

Reply:
It has to do with what this class is for, but parameterizing it by the getter function leads to logical clashes. In brief, ScalarView's creator needs to know the structure of ScalarView::Sample because it constructs ScalarView by passing a function returning a ScalarView::Sample. On the other hand, to peek at the ScalarView::Sample type, one needs to instantiate ScalarView, which is not possible unless a proxy type is used for Sample, which does not appear to be a good practice. So we have the chicken and the egg problem.

It is indeed possible to move Sample out of ScalarView, but since it is all about the coding style I am not sure if it is the right thing to do. ScalarView::Sample is logically the ScalarView's representation of samples, and ScalarView simply represents trials using a certain type while hiding the way those values are produced - the getter function. That's why I thought the most logical solution was to erase the getter's type. It is also most convenient for Python exports I think.